DOC: how to document examples using RNG and also self-contained bits (#18023)
scipy/scipy: doc/source/dev/contributor/rendering_documentation.rst

        Some text that follows the list.

Self-contained examples
~~~~~~~~~~~~~~~~~~~~~~~
All examples must be self-contained. It means all imports must be explicit,
the data used must be defined in the examples, etc.

    Yes::

        >>> import numpy as np
        >>> rng = np.random.default_rng()

    No::

        >>> rng = np.random.default_rng()

> The scope is defined

Reviewer: I don't understand this part about scope. I do understand the part about interspersing code and explanations. If they refer to the same idea, maybe the part about scope can be removed. If not, "i.e." is probably not appropriate.

Reviewer: I think that can be removed from here and clarified above, where each example is said to be self-contained. I'll make a suggestion.

per docstrings. i.e. for a given function's docstrings, it's possible,
and a good thing, to separate code with explanations. Note that blank lines
surrounding all block codes are required.

    Yes::

        Some initial text

        >>> import numpy as np
        >>> rng = np.random.default_rng()

        This is some explanation

        >>> rng.random(10)

Examples and randomness
~~~~~~~~~~~~~~~~~~~~~~~

> Examples are verified in the CI and output is compared. The main goal

Reviewer: ```suggestion
Examples are executed in the CI and the output is compared against the provided reference. The main goal
```

> is to ensure that the API and usage are consistent. Doctests are not unittests.

Reviewer: ```suggestion
is to ensure that the API and usage are consistent. Doctests are not meant to be used as unittests.
```
I'm not sure I understand "ensure that the API and usage". If I do, maybe this could be extended with something about: "a failure warns us that *the example* may need to be adjusted".

Reviewer: I will make a suggestion then.

> In case a random number generator is needed, `np.random.Generator` must be

Reviewer: Would it be wrong to state that `np.random.default_rng` must be used to generate one? This would be more obvious from the example below.

used.

    Yes::

        >>> import numpy as np
        >>> rng = np.random.default_rng()
        >>> sample = rng.random(10)

    Yes::

        >>> import numpy as np
        >>> rng = np.random.default_rng(102524723947864966825913730119128190984)
        >>> sample = rng.random(10)

    No::

        >>> import numpy as np
        >>> sample = np.random.random(10)

> Seeding the generator object is optional. Although they must not be common

Reviewer: ```suggestion
Seeding the generator object is optional. If a seed is used, avoid common numbers and
```

> numbers but instead be generated with: `np.random.SeedSequence().entropy`.

Reviewer: ```suggestion
instead generate a seed with: `np.random.SeedSequence().entropy`.
```

If no seed is provided, a default

> value is used when doctests are executed. In any case, the rendered

Reviewer: This would be a good place to indicate what that value is.

Reviewer: If it exists, I think it's still good to document it

> documentation does not show seed on purpose. The hope is that people do not

Reviewer: ```suggestion
documentation does not show the seed on purpose. The intent is to discourage users from
```

> copy/paste seeds in their code and are making an informed decision about

Reviewer: ```suggestion
copy/pasting seeds in their code and instead make an explicit decision about
```
(we can't really control whether it is "informed"!

        Some text that follows the list.

Self-contained examples
~~~~~~~~~~~~~~~~~~~~~~~

> All examples must be self-contained. It means all imports must be explicit,
> the data used must be defined in the examples, etc.

Reviewer: ```suggestion
Each "Example" section must be self-contained. This means that all imports
must be explicit, the data used must be defined, and the code should "just
work" when copy-pasted into a fresh Python interpreter.
```

    Yes::

        >>> import numpy as np
        >>> rng = np.random.default_rng()

    No::

        >>> rng = np.random.default_rng()

> The variable scope is defined
> per docstrings. i.e. for a given function's docstrings. It means variables
> cannot be shared between functions from the same file. What is possible,
> and recommended, is to separate code within a single example block with
> explanations. Note that blank lines surrounding all block codes are required.

Reviewer: ```suggestion
What is possible (and recommended) is to intersperse blocks of code with
explanations. Blank lines must separate each code block from the explanatory
text.
```

    Yes::

        Some initial text

        >>> import numpy as np
        >>> rng = np.random.default_rng()

        This is some explanation

        >>> rng.random(10)

Examples and randomness
~~~~~~~~~~~~~~~~~~~~~~~

> Examples are executed in the CI and the output is compared against the provided
> reference. The main goal is to ensure that the API is correctly documented and
> that we are not using outdated syntax. Doctests are not meant to be used as
> unittests.

Reviewer: Doctests can't directly verify that the "API is correctly documented", though. It can only verify that the example itself is working as intended. How about:
```suggestion
In the continuous integration (CI) suite, examples are executed and the output
is compared against the provided reference. The main goal is to ensure that
the *example* is correct; a failure warns us that the example may need to be
adjusted (e.g. because the API has changed since it was written).
Doctests are not meant to be used as unit tests of underlying implementation.
```

In case a random number generator is needed, `np.random.Generator` must be

> used. The canonical way to create a NumPy generator is to use

Reviewer: ```suggestion
used. The canonical way to create a NumPy ``Generator`` is to use
```

> `np.random.default_rng()`.

Reviewer: I think the point is to use the function `np.random.default_rng`, not necessarily the invocation `np.random.default_rng()`.

```suggestion
`np.random.default_rng`.
```
(hoping this will link with intersphinx? If not, `numpy.random.default_rng`.)

    Yes::

        >>> import numpy as np
        >>> rng = np.random.default_rng()
        >>> sample = rng.random(10)

    Yes::

        >>> import numpy as np
        >>> rng = np.random.default_rng(102524723947864966825913730119128190984)
        >>> sample = rng.random(10)

    No::

        >>> import numpy as np
        >>> sample = np.random.random(10)

Seeding the generator object is optional. If a seed is used, avoid common numbers and

> instead generate a seed with: `np.random.SeedSequence().entropy`.

Reviewer: ```suggestion
instead generate a seed with ``np.random.SeedSequence().entropy``.
```

If no seed is provided, the default value

> `1638083107694713882823079058616272161`

Reviewer: ```suggestion
``1638083107694713882823079058616272161``
```

> is used when doctests are executed. In any case, the rendered

Reviewer: ```suggestion
is used when doctests are executed. In either case, the rendered
```

> documentation does not show the seed on purpose. The intent is to discourage users from

Reviewer: ```suggestion
documentation will not show the seed. The intent is to discourage users from
```
("on purpose" is implied by "the intent")

> copy/pasting seeds in their code and instead make an explicit decision about
> the use of a seed in their program.

Reviewer: Last thing - what do you think about something like this?
```suggestion
copy/pasting seeds in their code and instead make an explicit decision about
the use of a seed in their program. The consequence is that users cannot
reproduce the results of the example exactly, so examples using random data
should not refer to precise numerical values based on random data or rely on
them to make their point.
```

Author: Fine with me 👍

        Some text that follows the list.

Self-contained examples
~~~~~~~~~~~~~~~~~~~~~~~

> Each "Example" section must be self-contained. This means that all imports

Reviewer 2: Optional, but may be worth mentioning
```suggestion
Each "Example" section (both in docstrings and general documentation) must be self-contained. This means that all imports
```
